storm-server/src/main/java/org/apache/storm/scheduler/utils/ArtifactoryConfigLoader.java

```diff
@@ -0,0 +1,395 @@
+import org.apache.storm.utils.Utils;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.SafeConstructor;
+
+/**
+ * A dynamic loader that can load scheduler configurations for user resource guarantees from Artifactory (an artifact repository manager).
+ */
+public class ArtifactoryConfigLoader implements IConfigLoader {
+    protected static final String LOCAL_ARTIFACT_DIR = "scheduler_artifacts";
+    static final String cacheFilename = "latest.yaml";
+    private static final String DEFAULT_ARTIFACTORY_BASE_DIRECTORY = "/artifactory";
+    private static final int DEFAULT_POLLTIME_SECS = 600;
+    private static final int DEFAULT_TIMEOUT_SECS = 10;
+    private static final String ARTIFACTORY_SCHEME_PREFIX = "artifactory+";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ArtifactoryConfigLoader.class);
+
+    @SuppressWarnings("rawtypes")
+    private Map loaderParams;
+    private int artifactoryPollTimeSecs = DEFAULT_POLLTIME_SECS;
+    private boolean cacheInitialized = false;
+    // Location of the file in the artifactory archive.  Also used to name file in cache.
+    private String localCacheDir;
+    private String baseDirectory = DEFAULT_ARTIFACTORY_BASE_DIRECTORY;
+    private int lastReturnedTime = 0;
+    private int timeoutSeconds = DEFAULT_TIMEOUT_SECS;
+    private Map lastReturnedValue;
+    private URI targetURI = null;
+    private JSONParser jsonParser;
+    private String scheme;
+
+    public ArtifactoryConfigLoader(Map loaderParams) {
+        this.loaderParams = loaderParams;
+        String thisTimeout = (String) loaderParams.get(ConfigLoaderConfiguration.SCHEDULER_CONFIG_LOADER_TIMEOUT_SECS);
+        if (thisTimeout != null) {
+            timeoutSeconds = Integer.parseInt(thisTimeout);
+        }
+        String thisPollTime = (String) loaderParams.get(ConfigLoaderConfiguration.SCHEDULER_CONFIG_LOADER_POLLTIME_SECS);
+        if (thisPollTime != null) {
+            artifactoryPollTimeSecs = Integer.parseInt(thisPollTime);
+        }
+        String thisBase = (String) loaderParams.get(ConfigLoaderConfiguration.ARTIFACTORY_BASE_DIRECTORY);
+        if (thisBase != null) {
+            baseDirectory = thisBase;
+        }
+        String uriString = (String) loaderParams.get(ConfigLoaderConfiguration.SCHEDULER_CONFIG_LOADER_URI);
+        if (uriString == null) {
+            LOG.error("No URI defined in {} configuration.", ConfigLoaderConfiguration.SCHEDULER_CONFIG_LOADER_URI);
+        } else {
+            try {
+                targetURI = new URI(uriString);
+                scheme = targetURI.getScheme().substring(ARTIFACTORY_SCHEME_PREFIX.length());
+            } catch(URISyntaxException e) {
+                LOG.error("Failed to parse uri={}", uriString);
```

Comment:
I think we want to throw an exception here.  I don't think we can or should recover from a bad URI.

Reply:
My movitation here is that if the ConfigLoader doesn't work correctly, we will just return null. Then [scheduler ](https://github.com/apache/storm/pull/2199/files#diff-7f8f870479c8f56254f940ea9fc70c92R65)will fall back to use "multitenant-scheduler.xml" file, or the configuration from storm.yaml. But I'm not sure if it's good

Reply:
I personally would prefer to see it fail fast here and not come up if the URI is not formatted correctly.